### PR TITLE
NIFI-3497 ScanAttribute Tagging

### PR DESCRIPTION
**Pull request by joetrite**
Adding tagging to ScanAttribute.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java [hunk 1]
```diff
@@ -97,13 +107,23 @@
             .addValidator(StandardValidators.createRegexValidator(0, 1, false))
             .defaultValue(null)
             .build();
+    public static final PropertyDescriptor DICTIONARY_ENTRY_METADATA_DEMARCATOR = new PropertyDescriptor.Builder()
+            .name("Dictionary Entry Metadata Demarcator")
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java]: The preferred code style is to use a "machine-readable" (aka no spaces) value for `name` and provide a human-formatted value for `displayName`.

Comment (joetrite) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java]: ok, cool.  i copied and pasted the existing properties which have not implemented displayName yet.  Do you want me to change it for the new properties that I added and leave the existing properties alone?

Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java]: Yes please. That recommendation was put in place after a lot of the code already existed, so what we are trying to enforce is that any new properties follow that model and people will go back and update existing properties as necessary. There have been many discussions about it, but one serious issue is that if a property name is changed (to make it more readable, rename it, fix a typo, etc.) the value stored in the `flow.xml.gz` file will not be picked up when the flow is read by NiFi and that data can be lost. So the combination of static `name` and malleable `displayName` solves this issue.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java [hunk 2]
```diff
@@ -97,13 +107,23 @@
...
+            .description("A single character used to demarcate the dictionary entry string between dictionary value and metadata.")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java]: If this is a single character, this validator should be a `StringLengthValidator` with a minimum and maximum of 1 character. See [here for an example](https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ExecuteProcess.java#L129-L135).

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java [hunk 3]
```diff
@@ -206,36 +243,60 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+
+        for (final Map.Entry<String, String> attribute : flowFile.getAttributes().entrySet()) {
+            if (attributePattern == null || attributePattern.matcher(attribute.getKey()).matches()) {
+                attributeNameMatches.add(attribute.getKey());
+
+                if (dictionary.containsKey(attribute.getValue())) {
+                    hitCounter++;
+                    dictionaryTermMatches.put("dictionary.hit." + hitCounter + ".attribute", attribute.getKey());
+                    dictionaryTermMatches.put("dictionary.hit." + hitCounter + ".term", attribute.getValue());
+                    dictionaryTermMatches.put("dictionary.hit." + hitCounter + ".metadata", dictionary.get(attribute.getValue()));
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java]: It looks like these statements could be refactored to a method as they are reused in the `matchAll` scenario.

**Comment by joewitt** [comment]
@joetrite it would be good to get a clean master, create a branch, put your changes on it, then submit that PR.  This one has a couple merge commits now and they're probably not needed (rebase would likely have done the trick).  Can you clean up the commit history for this PR?  Definitely appreciate the work you're putting in and don't want to put you through hoops but would like to keep the commit clean.

**Comment by joetrite** [comment]
created new pull request #1564 with single commit. closing this.  thanks @joewitt @alopresto for all the help and advise.

**Pull request closed without merging**